File: wear/app/src/main/java/com/gearan/watch/pairing/GearanPairingManager.kt
package com.gearan.watch.pairing

import com.gearan.watch.security.CryptoUtils
import com.gearan.watch.security.GearanSecureSession
import com.gearan.watch.util.GearanLog
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

/**
 * Watch-side pairing orchestrator: advertising -> handshake -> SAS display ->
 * user confirm -> secure -> save trusted -> initial sync signal.
 * 120 s timeout, full error mapping.
 */
class GearanPairingManager(private val scope: CoroutineScope) {
    private val machine = GearanPairingStateMachine()
    private val _state = MutableStateFlow(PairingState.IDLE)
    val state: StateFlow<PairingState> = _state
    private val _sas = MutableStateFlow<String?>(null)
    val sas: StateFlow<String?> = _sas
    private val _error = MutableStateFlow<PairingError?>(null)
    val error: StateFlow<PairingError?> = _error

    val secureSession = GearanSecureSession()
    private var timeoutJob: Job? = null

    private fun emit(s: PairingState) {
        if (s != _state.value) GearanLog.pair("state ${_state.value} -> $s")
        _state.value = s
    }

    fun startPairing(onAdvertise: () -> Unit) {
        reset()
        emit(machine.send(PairingEvent.StartPairing))
        secureSession.beginHandshake()
        onAdvertise()
        armTimeout()
    }

    fun onCentralConnected() {
        emit(machine.send(PairingEvent.LinkConnected))
        emit(machine.send(PairingEvent.LinkConnected)) // CONNECTING -> HANDSHAKING
    }

    fun onHandshakeMaterialReady(): String {
        val sas = secureSession.deriveSession()
        _sas.value = sas
        emit(machine.send(PairingEvent.HandshakeDone(sas)))
        return sas
    }

    fun confirmCode() {
        emit(machine.send(PairingEvent.UserConfirmed))
    }

    fun cancelPairing(reason: String = "user cancelled") {
        emit(machine.send(PairingEvent.UserCancelled))
        fail(PairingError("CANCELLED", reason))
    }

    fun onSessionSecured() = emit(machine.send(PairingEvent.SessionSecured))
    fun onTrustedSaved() = emit(machine.send(PairingEvent.TrustedSaved))
    fun onSyncDone() {
        emit(machine.send(PairingEvent.SyncDone))
        cancelTimeout()
    }

    fun onLinkLost() { emit(machine.send(PairingEvent.LinkLost)) }
    fun onReconnected() { emit(machine.send(PairingEvent.Reconnected)) }

    fun fail(e: PairingError) {
        _error.value = e
        GearanLog.pairWarn("pairing failed [${e.code}]: ${e.message}")
        emit(machine.send(PairingEvent.Failed(e)))
        cancelTimeout()
    }

    fun reset() {
        cancelTimeout()
        _sas.value = null
        _error.value = null
        secureSession.reset()
        machine.send(PairingEvent.Reset)
        emit(machine.state)
    }

    private fun armTimeout() {
        cancelTimeout()
        timeoutJob = scope.launch {
            delay(CryptoUtils.PAIRING_TIMEOUT_MS)
            if (_state.value != PairingState.PAIRED && _state.value != PairingState.CONNECTED) {
                fail(PairingError("TIMEOUT", "Pairing timed out after 120 s"))
            }
        }
    }

    private fun cancelTimeout() { timeoutJob?.cancel(); timeoutJob = null }
}

File: ios/Sources/GearanCore/GearanPairingManager.swift
import Foundation

/// Pairing lifecycle — identical state set to wear/GearanPairingStateMachine.
public enum PairingState: String, Equatable {
    case idle, scanning, watchFound, connecting, handshaking
    case waitingForUserConfirmation, securingConnection, savingTrustedDevice
    case initialSync, paired, reconnecting, connected, disconnected, failed
}

public struct PairingError: Error, Equatable {
    public var code: String; public var message: String
    public init(code: String, message: String) { self.code = code; self.message = message }
}

public enum PairingEvent {
    case startScanning, watchFound, connected, handshakeDone(sas: String)
    case userConfirmed, userCancelled, sessionSecured, trustedSaved, syncDone
    case linkLost, reconnected, failed(PairingError), reset
}

/// Pure state machine (no Bluetooth) — unit-tested.
public final class GearanPairingStateMachine {
    public private(set) var state: PairingState = .idle
    public init(_ initial: PairingState = .idle) { state = initial }
    @discardableResult
    public func send(_ e: PairingEvent) -> PairingState {
        switch state {
        case .idle:
            if case .startScanning = e { state = .scanning }
        case .scanning:
            switch e {
            case .watchFound: state = .watchFound
            case .connected: state = .connecting
            case .failed: state = .failed
            case .reset: state = .idle
            default: break
            }
        case .watchFound:
            switch e {
            case .connected: state = .connecting
            case .failed: state = .failed
            case .reset: state = .idle
            default: break
            }
        case .connecting:
            switch e {
            case .connected: state = .handshaking
            case .handshakeDone: state = .waitingForUserConfirmation
            case .failed: state = .failed
            case .reset: state = .idle
            default: break
            }
        case .handshaking:
            switch e {
            case .handshakeDone: state = .waitingForUserConfirmation
            case .failed: state = .failed
            case .reset: state = .idle
            default: break
            }
        case .waitingForUserConfirmation:
            switch e {
            case .userConfirmed: state = .securingConnection
            case .userCancelled: state = .disconnected
            case .failed: state = .failed
            case .reset: state = .idle
            default: break
            }
        case .securingConnection:
            switch e {
            case .sessionSecured: state = .savingTrustedDevice
            case .failed: state = .failed
            default: break
            }
        case .savingTrustedDevice:
            switch e {
            case .trustedSaved: state = .initialSync
            case .failed: state = .failed
            default: break
            }
        case .initialSync:
            switch e {
            case .syncDone: state = .paired
            case .failed: state = .failed
            default: break
            }
        case .paired:
            switch e {
            case .reconnected: state = .connected
            case .linkLost: state = .reconnecting
            case .reset: state = .idle
            default: break
            }
        case .reconnecting:
            switch e {
            case .reconnected: state = .connected
            case .failed: state = .failed
            case .reset: state = .idle
            default: break
            }
        case .connected:
            switch e {
            case .linkLost: state = .reconnecting
            case .reset: state = .idle
            default: break
            }
        case .disconnected:
            switch e {
            case .startScanning: state = .scanning
            case .reset: state = .idle
            default: break
            }
        case .failed:
            switch e {
            case .reset: state = .idle
            case .startScanning: state = .scanning
            default: break
            }
        }
        return state
    }
}
